### PR TITLE
Loaded only the URL relations live routes need under lazy routing

### PR DESCRIPTION
**Pull request by vershwal**
ref https://linear.app/ghost/issue/HKG-1870

Lazy URL routing resolves a resource's URL by evaluating each router's NQL
filter against the in-memory resource, so any resource passed to
`getUrlForResource` must already have the relations those filters reference
(e.g. `tags.slug`, `authors.slug`) loaded — otherwise it 404s as a "thin"
resource. The original lazy-404 fix (#27938) handled this by force-loading a
blanket `tags` + `authors` on a wide set of post/page reads. Some of those
loads ran unconditionally (email, comments), which added query noise and a
measurable performance cost even on sites that never use lazy routing.

This PR makes that loading precise and routes it all through one helper, ahead
of enabling compare-mode rollout (TryGhost/Zuul#1133).

## Changes

- **Added `getRequiredRelations()` to the lazy URL service + facade.** It scans
  the registered router filters and returns only the subset of `tags`/`authors`
  the live routes actually reference (`[]` when none do), cached and invalidated
  when routers change. The facade returns `[]` when there is no lazy backend, so
  eager and non-lazy sites compute nothing.
- **Switched the gated API call sites to the helper.** `search-index`,
  `search-index-public`, and the posts/pages input serializers now load only the
  referenced relations on the `?fields=url` path instead of a blanket
  `tags+authors`, and the early-return bypass when a caller passes both
  `?fields=url` and `?include=` is fixed.
- **Routed the unconditional email/comments callers through the helper via
  dependency injection.** `email-renderer`, `batch-sending-service`, and
  `email-controller` take an injected `getRequiredUrlRelations` wired by the
  email-service composition root; `comments-service` uses the `urlService` it was
  already handed. Each path keeps its own render relations (`posts_meta`,
  `authors`) intact.

## Safety / rollout

- **Eager (default, all self-hosters + 80% of Pro):** `getRequiredRelations()`
  returns `[]`, so no extra relations are loaded. Eager resolves URLs by id and
  never touches a resource's relations; anything needed for rendering is fetched
  via `getLazyRelation`. This removes the unconditional loads #27938 added and
  restores pre-#27938 query behavior.
- **Lazy compare mode (TryGhost/Zuul#1133, 20% of Pro):** eager remains the
  canonical, returned answer; lazy runs in the background. We load exactly the
  relations lazy needs to evaluate filters, so any edge case surfaces as a
  logged comparison mismatch rather than a user-facing 404.

## Test plan

- [x] `getRequiredRelations()` unit tests (lazy service + facade)
- [x] Updated posts/comments input-serializer tests to stub the helper
- [x] email-controller / batch-sending / comments-service unit tests pass
- [x] `tsc --noEmit` + eslint clean
- [ ] Watch compare-mode mismatch logs after Zuul#1133 ships at 20%

### CODE REVIEW

File: ghost/core/core/server/services/url/url-service-facade.ts [hunk 1]
```diff
@@ -155,6 +156,15 @@ export class UrlServiceFacade {
...
     }
 
+    // Returns [] when there is no lazy backend: eager looks URLs up by id and
+    // never touches a resource's relations.
+    getRequiredRelations(): string[] {
+        if (this.lazyUrlService) {
+            return this.lazyUrlService.getRequiredRelations();
+        }
+        return [];
+    }
```
[the remark this answers is not in the document]
Comment (vershwal) [reply ghost/core/core/server/services/url/url-service-facade.ts:166]: `getRequiredRelations()` isn't part of the fire-and-forget comparison path — unlike `_compare`/`_runLazyHook`, which wrap the background `getUrlForResource`/`resolveUrl` calls that do real DB work and run via `setImmediate`/`void`. This is a synchronous, foreground decision used to pick `withRelated` before the query runs, and the lazy implementation is a pure in-memory regex scan over the registered router filters (no I/O, guarded against `null` filters), so it can't realistically throw. Wrapping it to fall back to `[]` would also be counterproductive in primary lazy mode (`isLazy()`), where silently returning `[]` would under-load relations and cause real user-facing thin-resource 404s with no signal — the opposite of what we'd want. Leaving as-is.

File: ghost/core/core/server/services/url/lazy-url-service.ts [hunk 2]
```diff
@@ -72,6 +75,27 @@ export class LazyUrlService implements LazyUrlServiceBackend {
...
+            if (/\btags?\b/.test(config.filter) || /\bprimary_tag\b/.test(config.filter)) {
+                required.add('tags');
+            }
+            if (/\bauthors?\b/.test(config.filter) || /\bprimary_author\b/.test(config.filter)) {
+                required.add('authors');
+            }
+        }
+        this.requiredRelations = [...required];
+        return this.requiredRelations;
     }
```
[the remark this answers is not in the document]
Comment (vershwal) [reply ghost/core/core/server/services/url/lazy-url-service.ts:108]: Fixed here dfd2fcca75aa78fa0a0b3b2e6c318d4b0c108451

File: ghost/core/core/server/services/url/lazy-url-service.ts [hunk 3]
```diff
@@ -72,6 +75,27 @@ export class LazyUrlService implements LazyUrlServiceBackend {
...
+            if (/\btags?\b/.test(config.filter) || /\bprimary_tag\b/.test(config.filter)) {
+                required.add('tags');
+            }
+            if (/\bauthors?\b/.test(config.filter) || /\bprimary_author\b/.test(config.filter)) {
+                required.add('authors');
+            }
+        }
+        this.requiredRelations = [...required];
+        return [...this.requiredRelations];
     }
```
[the remark this answers is not in the document]
Comment (vershwal) [reply ghost/core/core/server/services/url/lazy-url-service.ts:108]: Good catch — valid. `getRequiredRelations()` was only scanning filters, but `replacePermalink` runs the permalink template too, so a `:primary_tag`/`:primary_author` permalink with no matching filter would load nothing and generate a wrong URL. The old blanket `tags+authors` load masked this; narrowing exposed it. Fixed by also scanning `config.permalink` for `primary_tag`/`primary_author` (permalink-only, to avoid re-broadening on taxonomy `/tag/`,`/author/` segments). Added tests for the no-filter permalink case.
